src/NanopassSharp/Pass.cs

```diff
@@ -0,0 +1,71 @@
+﻿using System.Collections.Generic;
+
+namespace NanopassSharp;
+
+/// <summary>
+/// A single compiler pass.
+/// </summary>
+/// <param name="Name">The name of the pass.</param>
+/// <param name="Documentation">The pass' corresponding documentation.</param>
+/// <param name="Transformations">The transformations the pass applies to its nodes.</param>
+/// <param name="Tree">The original node tree of the pass.</param>
+public sealed record class Pass
+(
+    string Name,
+    string? Documentation,
+    Pass Previous,
+    IList<ITransformation> Transformations,
+    // Should this be immutable?
+    // Should the tree be the tree resulting from applying the pass' transformations?
+    // Should it be created immediately by the language-specific parser
+    // or be based on the previous pass?
+    TypeNodeTree Tree
+);
+
+/// <summary>
+/// A tree of nodes representing types which are the input and output of a pass.
+/// </summary>
+/// <param name="Root">The root node.</param>
+/// <param name="Nodes">The nodes of the tree.</param>
+public sealed record class TypeNodeTree
```

Comment:
I'm not a fan of this name, it doesn't immediately yell that this is the node hierarchy.